### DISCUSSION TITLE
Support TOTP with normal Users and Admins

### DISCUSSION THREAD

**Discussion by niklasarnitz** [pocketbase/pocketbase#1208] [Q&A]
Hi,

It would be awesome, if pocketbase could support TOTP 2FA auth.

**Comment by ganigeorgiev** [comment]
I don't think we can support this in a generic way since most of the times user will expect the 2fa code to be send via sms or similar and this is different for each provider.

It also may not be necessary in the future because of the eventual webauthn integration (https://github.com/pocketbase/pocketbase/issues/406).

**Reply by niklasarnitz** [reply]
I tend to digress.
There is an industry standard for 2FA authentication: TOTP (https://en.m.wikipedia.org/wiki/Time-based_one-time_password).
And it is widely supported by many apps.

**Reply by niklasarnitz** [reply]
And passkeys don’t solve that problem, because, the way it’s looking right now they will only be available to certain devices.
TOTP works without device dependencies.

**Reply by ganigeorgiev** [reply]
@niklasarnitz

> There is an industry standard for 2FA authentication

I meant not the code itself but the way how you'll send it to the user since providers use different SMS apis.

> TOTP works without device dependencies.

Maybe I'm misunderstanding the request. Could you provide an example/expected flow?

**Reply by niklasarnitz** [reply]
TOTP works completely offline and without API endpoints.
A flow for the initial setup is:
- The Webapp shows a QR Code to the user
- This QR Code contains the OTP Secret Key, issuer and key validity period
- This QR Code is then scanned by the user in an Authenticator app like Authy or Google Authenticator
- The authenticator app then can generate TOTP codes completely offline via algorithms that use the data given by the qr code to generate 6 Digit number codes
- On the server, the same algorithms are executed which then can check if the entered code of  the user matches the one generated by the server.

**Reply by niklasarnitz** [reply]
Here is a nice explanation of the working principle: https://www.twilio.com/docs/glossary/totp

**Reply by Ponkhy** [reply]
I agree with @niklasarnitz it's pretty much a standard and the user can use whatever TOTP/HOTP app they want. There is no code that will directly send to the user, rather the user generate it themself. In my case it would be Bitwarden, but there is also the Microsoft, Google Authenticator and many more. It's also quite easy to write a authentication app or implement a 2FA function based on TOTP/HOTP into systems.

**Reply by ganigeorgiev** [reply]
@niklasarnitz Thank you for providing more details. I confused it with otp, which I suppose is the sms code case.

I personally don't have a use case for this and I think most users will find the flow complicated, and therefore at least for v1 I'm not planning to work on it, but that doesn't mean that you can't implement it on your own if you need it and attach it as "plugin" to the `core.App` instance (in v0.10 or v0.11 there will be also an option to bind to the app hooks with JS).

**Reply by 1l0** [reply]
@ganigeorgiev seems https://github.com/pquerna/otp is a great reference for TOTP, I've tested and it just works.

**Reply by HeyElix** [reply]
@ganigeorgiev 

Hi, I think TOTP is a must, developers are very familiar with Google Authenticator  bindings and the process is not complicated at all, it's a must. I appreciate your contribution to pocketbase, but drop your bias and listen more to what real users feel.       Google Authenticator can be an optional, not mandatory binding, but for backend admin users it is a must because I value security very highly and security is a priority.


1. **Enhanced Security**: TOTP provides an additional layer of security for user accounts. Even if an attacker obtains a user's static password, they cannot access the account without the one-time password generated by TOTP.

2. **Prevention of Replay Attacks**: Since TOTP passwords are time-based and a new password is generated every 30 or 60 seconds, any intercepted one-time password cannot be used in the next time window for a replay attack.

3. **Cross-Platform Support**: TOTP Authenticator apps support multiple platforms, including iOS and Android, allowing users to generate and use one-time passwords on different devices, which increases flexibility.

4. **Ease of Backup and Recovery**: Some TOTP Authenticator apps offer cloud synchronization features, allowing users to back up their 2FA accounts to the cloud, making it easy to recover when switching devices and reducing the risk of losing access.

5. **No Need for Network Connection**: Generating TOTP passwords does not require a network connection, so users can generate one-time passwords for authentication even when offline, enhancing usability in various environments.

6. **Standardization and Compatibility**: TOTP follows an open standard (RFC 6238), ensuring compatibility between different services and applications. Users can manage the 2FA needs of multiple services with one Authenticator app, simplifying the management process.

7. **User Experience**: Compared to traditional SMS verification codes, TOTP offers a faster and more convenient user experience. Users do not have to wait for an SMS to arrive and do not have to worry about the instability of SMS services.

8. **Cost-Effectiveness**: For service providers, TOTP reduces the cost of sending SMS verification codes, while also providing a secure solution for users without additional expenses.

In summary, using a TOTP Authenticator significantly increases the security of accounts, prevents unauthorized access, and provides a convenient, flexible, and cost-effective two-factor authentication solution.

**Reply by ganigeorgiev** [reply]
@thevincentlee There is no need for such passive-aggressive tone and the above doesn't really add anything to the discussion (_and if it was AI generated please stop doing it_).

I have no "bias" against TOTP and we may implement it in the future but I have already a lot on my plate and there are many other critical tasks with higher priority (see also https://github.com/pocketbase/pocketbase/discussions/4355).

PocketBase is a personal/hobby project that I've created for my own needs (see [Presentator #183](https://github.com/presentator/presentator/issues/183)) and I usually prioritize tasks based on whether they are a bug, blocker or something that I need in my projects.

TOTP is neither of those. As mentioned in the other comments, user contributions are welcomed. This feature specifically doesn't necessary needs to be part of the core package and can be implemented as "plugin"/library. Feel free to create a POC implementation and I may consider prioritizing it after the refactoring.

**Reply by HeyElix** [reply]
Sorry for the misunderstanding I didn't mean it that way, PocketBase is great and I use it all the time, I just want PocketBase to get better, nothing personal, I understand you are busy alone and have higher priority tasks.  ❤️

**Comment by dxvid-pts** [comment]
Implementing this feature would be highly beneficial, not only for end users but also for administrators in terms of securing the administrative interface.

**Comment by gaby** [comment]
What's there any update to this? More, and more apps are starting to use Pocketbase as a backend the lack of TOTP will become a serious issue if one of these instances gets compromise.

**Comment by miztroh** [comment]
@ganigeorgiev Could you provide a bit more clarity on this?  I'm not understanding from the documentation where I could insert something custom into the admin login process.  Here's how I'm imagining something like this would work:

1. user visits normal admin login page
2. user provides email, password, clicks login
3. custom integration determines whether a user has configured TOTP (perhaps with custom collection)
4. if user has not configured TOTP, proceed with normal login process
5. if user has configured TOTP, prompt for TOTP code.  If failure, redirect to login page.  If success, proceed with normal login process.

On top of the above, there'd have to be some mechanism for adding admin UI that allows a user to perform the initial TOTP setup.

I know you mentioned you don't have a use case for this, but there needs to be something to protect admin accounts from brute force attacks.  I'm happy to pursue a custom integration, but it's not clear from the docs how this would even be possible from an admin user perspective.

Thank you!

**Reply by ganigeorgiev** [reply]
Currently the admin and regular users are 2 different entities (this will change with the [planned refactoring](https://github.com/pocketbase/pocketbase/discussions/4355)).

For regular user auth you can explore the available [Record Auth hooks](https://pocketbase.io/docs/go-event-hooks/#record-auth-api-hooks).

Similarly, for the admin users you could check [Admin Auth Hooks](https://pocketbase.io/docs/go-event-hooks/#admin-auth-api-hooks).


> On top of the above, there'd have to be some mechanism for adding admin UI that allows a user to perform the initial TOTP setup.

The Admin UI is not extendable at the moment but that doesn't mean that you can't register a custom route that will return your own html.

For an example how to extend the auth you could check the Gist linked in https://github.com/pocketbase/pocketbase/issues/119#issuecomment-1881906233 (it is for OTP but I guess something similar could be applied for TOTP).

>  but there needs to be something to protect admin accounts from brute force attacks

An alternative if you want to prevent admin login attempts from malicious users could be to register register a global middleware with [Go](https://pocketbase.io/docs/go-routing/#middlewares) or [JS](https://pocketbase.io/docs/js-routing/#middlewares) that will reject `/api/admins/*` requests from non-whitelisted IPs.

**Comment by greendesertsnow** [comment]
Hey there,
When we serve pocketbase out there, the admin panel is/will eventually be exposed to brute force attacks with a username/password combination. Enabling 2FA through an OTP mail can boost the security as PB already has SMTP.

**Reply by Tonku** [reply]
you can use cloudflare to rate limit admin url

**Comment by ganigeorgiev** [comment]
I'm closing this discussion. OTP (email code) and MFA is already implemented as part of the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) and it will be available with the next major release (there are no ETAs). TOTP support for now remains unplanned but I may reconsider it once the other more critical/blocker tasks are completed (and again, it doesn't need to be part of the main repo and it can be implemented externally as "plugin" package via the existing hooks).
